### PR TITLE
rte: controller: push down machineconfig deletion support into library packages

### PR DESCRIPTION
**Pull request by ffromani**
Push down the logic to manage the deletion of machineconfig into library packages, streamlining the controller.

This PR is meant to do code movement and reorganize the packages, no intended changes in behavior

### CODE REVIEW

**Comment by Tal-or** [comment]
Lets add another commit which actually wiring this function so we can see that it works and tests are passing correctly.

**Comment by ffromani** [comment]
/hold

need to remove the demo commit

**Comment by Tal-or** [comment]
> /hold
> 
> need to remove the demo commit

What are we waiting for? if it's working let's keep it. One less thing to worry for

**Comment by ffromani** [comment]
> > /hold
> > need to remove the demo commit
> 
> What are we waiting for? if it's working let's keep it. One less thing to worry for

Because I'd introduce the new API in the new HCP-related flow first. There are too many changes in flight already (for the resources we have) and I want to contain the interference radius but still keep improving the codebase.
It's a tradeoff.

**Comment by Tal-or** [comment]
> > > /hold
> > > need to remove the demo commit
> > 
> > 
> > What are we waiting for? if it's working let's keep it. One less thing to worry for
> 
> Because I'd introduce the new API in the new HCP-related flow first. There are too many changes in flight already (for the resources we have) and I want to contain the interference radius but still keep improving the codebase. It's a tradeoff.

Ack. so shall we unhold #1025? I kept the hold only because I thought we want this change, but it won't affect anyway

**Comment by ffromani** [comment]
> > > > /hold
> > > > need to remove the demo commit
> > > 
> > > 
> > > What are we waiting for? if it's working let's keep it. One less thing to worry for
> > 
> > 
> > Because I'd introduce the new API in the new HCP-related flow first. There are too many changes in flight already (for the resources we have) and I want to contain the interference radius but still keep improving the codebase. It's a tradeoff.
> 
> Ack. so shall we unhold #1025? I kept the hold only because I thought we want this change, but it won't affect anyway since it's not wired.

yes, I'd unhold #1025, then I'll rebase this one on top and port only the new flow to the new API. We will add other improvements later.

**Comment by Tal-or** [comment]
CI passed. you may remove the demo commit, i'll approve

File: pkg/objectstate/objectstate.go [hunk 1]
```diff
@@ -37,3 +37,7 @@ func (obst ObjectState) IsNotFoundError() bool {
 	}
 	return apierrors.IsNotFound(obst.Error)
 }
+
+func (obst ObjectState) IsCreateOrUpdate() bool {
```
Comment (ffromani) [inline pkg/objectstate/objectstate.go:41]: needs test also to catch the `non-nil interface point-to-nil` (https://trstringer.com/go-nil-interface-and-interface-with-nil-concrete-value/)

**Review by ffromani — COMMENTED** [review]
/hold

some tests are still missing

File: pkg/objectstate/objectstate.go (hunk 1 above)
Comment (ffromani) [reply pkg/objectstate/objectstate.go:43]: solved in the last upload

File: pkg/apply/apply.go [hunk 2]
```diff
@@ -27,24 +27,44 @@ import (
...
 func ApplyObject(ctx context.Context, cli k8sclient.Client, objState objectstate.ObjectState) (k8sclient.Object, bool, error) {
+	// both `objState.Existing` and `objState.Desired` are assumed not-nil
 	objDesc, _ := describeObject(objState.Desired)
 
 	if objState.IsNotFoundError() {
 		klog.InfoS("creating", "object", objDesc)
 		err := cli.Create(ctx, objState.Desired)
 		if err != nil {
+			err = errors.Wrapf(err, "could not create object %s", objDesc)
 			return nil, false, err
```
Comment (Tal-or) [inline pkg/apply/apply.go:68]: `return nil, false, fmt.Errorf("could not create object %s %w", objDesc, err)`

File: pkg/apply/apply.go [hunk 3]
```diff
@@ -27,24 +27,44 @@ import (
...
+func ApplyState(ctx context.Context, cli k8sclient.Client, objState objectstate.ObjectState) (k8sclient.Object, bool, error) {
+	if !objState.IsCreateOrUpdate() { // delete requested
+		// `objState.Desired` should be assumed `nil` in this branch and never be used
+		if objState.IsNotFoundError() { // and nothing found: nothing to do
+			return nil, false, nil
+		}
+		if objState.Existing == nil {
+			err := fmt.Errorf("inconsistent internal state: delete desired, but nil existing object")
+			klog.ErrorS(err, "applyState")
+			return nil, false, err
```
Comment (Tal-or) [inline pkg/apply/apply.go:43]:
```
return nil, false, fmt.Errorf("inconsistent internal state: delete desired, but nil existing object")
```

File: pkg/apply/apply.go [hunk 4]
```diff
@@ -27,24 +27,44 @@ import (
...
+		}
+		objDesc, _ := describeObject(objState.Existing)
+		klog.InfoS("deleting", "object", objDesc)
+
+		err := cli.Delete(ctx, objState.Existing)
+		if err != nil {
+			err = errors.Wrapf(err, "could not delete object %s", objDesc)
+			return nil, false, err
```
Comment (Tal-or) [inline pkg/apply/apply.go:51]:
```
return nil, false, fmt.Errorf("could not delete object %s %w", objDesc, err)
```

**Review by Tal-or — COMMENTED** [review]
LGTM
small comments

**Comment by ffromani** [comment]
I'll remove all `errors.Wrapf` in one go in a followup PR:
```
controllers/kubeletconfig_controller.go:200:			return nil, errors.Wrapf(err, "Failed to set controller reference to %s %s", objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/kubeletconfig_controller.go:203:			return nil, errors.Wrapf(err, "could not create %s", objState.Desired.GetObjectKind().GroupVersionKind().String())
controllers/numaresourcesoperator_controller.go:186:		return false, errors.Wrapf(err, "could not update status for object %s", client.ObjectKeyFromObject(instance))
controllers/numaresourcesoperator_controller.go:206:		return true, ctrl.Result{}, status.ConditionDegraded, errors.Wrapf(err, "FailedAPISync")
controllers/numaresourcesoperator_controller.go:220:		return true, ctrl.Result{}, status.ConditionDegraded, errors.Wrapf(err, "failed to sync machine configs")
controllers/numaresourcesoperator_controller.go:241:		return true, ctrl.Result{}, status.ConditionDegraded, errors.Wrapf(err, "FailedRTESync")
controllers/numaresourcesoperator_controller.go:313:			err = errors.Wrapf(err2, "could not create %s", objState.Desired.GetObjectKind().GroupVersionKind().String())
controllers/numaresourcesoperator_controller.go:341:				err = errors.Wrapf(err2, "failed to set controller reference to %s %s", objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/numaresourcesoperator_controller.go:479:			return nil, errors.Wrapf(err, "failed to update (%s) %s/%s", objState.Desired.GetObjectKind().GroupVersionKind(), objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/numaresourcesoperator_controller.go:483:			return nil, errors.Wrapf(err, "failed to set controller reference to %s %s", objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/numaresourcesoperator_controller.go:487:			return nil, errors.Wrapf(err, "failed to apply (%s) %s/%s", objState.Desired.GetObjectKind().GroupVersionKind(), objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/numaresourcesscheduler_controller.go:149:		return ctrl.Result{}, status.ConditionDegraded, errors.Wrapf(err, "FailedSchedulerSync")
controllers/numaresourcesscheduler_controller.go:238:			return schedStatus, errors.Wrapf(err, "Failed to set controller reference to %s %s", objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/numaresourcesscheduler_controller.go:242:			return schedStatus, errors.Wrapf(err, "could not apply (%s) %s/%s", objState.Desired.GetObjectKind().GroupVersionKind(), objState.Desired.GetNamespace(), objState.Desired.GetName())
controllers/numaresourcesscheduler_controller.go:258:		return errors.Wrapf(err, "could not update status for object %s", client.ObjectKeyFromObject(sched))
pkg/apply/apply.go:50:			err = errors.Wrapf(err, "could not delete object %s", objDesc)
pkg/apply/apply.go:67:			err = errors.Wrapf(err, "could not create object %s", objDesc)
pkg/apply/apply.go:77:		return nil, false, errors.Wrapf(err, "could not merge object %s with existing", objDesc)
pkg/apply/apply.go:81:		return nil, false, errors.Wrapf(err, "could not compare object %s with existing", objDesc)
pkg/apply/apply.go:87:			return nil, updated, errors.Wrapf(err, "could not update object %s", objDesc)
pkg/hash/configmap.go:48:			return "", errors.Wrapf(err, "could not calculate ConfigMap %q hash", key.String())
```

**Comment by ffromani** [comment]
here: https://github.com/openshift-kni/numaresources-operator/pull/1039
